### PR TITLE
Rewrite most of napi_threadsafe_function

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

Supersedes https://github.com/oven-sh/bun/pull/15289

Fixes https://github.com/oven-sh/bun/issues/14664
Fixes https://github.com/oven-sh/bun/issues/12977
Fixes https://github.com/oven-sh/bun/issues/6599

### CODE REVIEW

**Comment by robobun** [comment]
@Jarred-Sumner, your commit 69c0a26eb350953ddec44ab3569c11a437ba6d25 is building: [`#6557`](https://buildkite.com/bun/bun/builds/6557)

File: src/napi/napi.zig [hunk 1]
```diff
@@ -1426,148 +1434,236 @@ pub const ThreadSafeFunction = struct {
...
-                },
-                else => {
-                    const slice = allocator.alloc(?*anyopaque, size) catch unreachable;
-                    return .{
-                        .sized = Channel(?*anyopaque, .Slice).init(slice),
-                    };
-                },
-            }
+        pub fn isBlocked(this: *const Queue) bool {
+            return this.max_queue_size > 0 and this.count.load(.monotonic) >= this.max_queue_size;
```
Comment (190n) [inline src/napi/napi.zig]: Change all atomic ops in ThreadsafeFunction from `monotonic` to `seq_cst`

File: src/napi/napi.zig [hunk 2]
```diff
@@ -1426,148 +1434,236 @@ pub const ThreadSafeFunction = struct {
...
-            if (!try this.channel.tryWriteItem(ctx)) {
-                return error.WouldBlock;
+            if (this.queue.isBlocked()) {
+                return .queue_full;
             }
         }
 
-        this.event_loop.enqueueTaskConcurrent(JSC.ConcurrentTask.createFrom(this));
+        this.lock.lock();
+        defer this.lock.unlock();
```
Comment (190n) [inline src/napi/napi.zig]: These lines should be at the top of the function. Waiting on a condition variable requires the mutex to already be held by the waiting thread, and the mutex is re-acquired inside wait() before wait() returns.

File: src/napi/napi.zig [hunk 3]
```diff
@@ -1426,148 +1434,236 @@ pub const ThreadSafeFunction = struct {
...
-        pub fn isClosed(this: *const @This()) bool {
-            return @atomicLoad(
-                bool,
-                switch (this.*) {
-                    .sized => &this.sized.is_closed,
-                    .unsized => &this.unsized.is_closed,
-                },
-                .seq_cst,
-            );
+        // This value will never change after initialization.
```
Comment (190n) [inline src/napi/napi.zig]:
```suggestion
        /// This value will never change after initialization. Zero means the size is unlimited.
```

File: src/napi/napi.zig [hunk 4]
```diff
@@ -1426,148 +1434,236 @@ pub const ThreadSafeFunction = struct {
...
+        max_queue_size: usize,
+
+        count: std.atomic.Value(u32) = std.atomic.Value(u32).init(0),
```
Comment (190n) [inline src/napi/napi.zig:1482]: Can we just use the count tracked by the fifo?

File: src/napi/napi.zig [hunk 5]
```diff
@@ -1426,148 +1434,236 @@ pub const ThreadSafeFunction = struct {
...
+                // we've already scheduled it to run
+            },
+        }
+    }
+
+    pub fn deinit(this: *ThreadSafeFunction) void {
         this.unref();
 
         if (this.finalizer.fun) |fun| {
             fun(this.event_loop.global, this.finalizer.data, this.ctx);
```
Comment (190n) [inline src/napi/napi.zig:1669]:
```suggestion
            const handle_scope = NapiHandleScope.open(this.env, false);
            defer if (handle_scope) |scope| scope.close(this.env);
            fun(this.event_loop.global, this.finalizer.data, this.ctx);
```

**Review by 190n — CHANGES_REQUESTED** [review]

**Pull request merged**